04 Strategy Library/999 Seasonality and Session Effect in Index Futures/03 Method.html

```diff
@@ -0,0 +1,124 @@
+    
+    
+    # Bullish Season: buy at 18:00, sell at 5:30; buy at 13:15, sell at 15:15
+    # Bearlish Season: buy at 18:00, sell at 5:30; buy at 13:15, sell at 14:00
+    
+    
+    # Night: 1800 - 530
+    # Afternoon: 1315 - 1400
+    # Close: 1400 - 1515
+    
+    # Night session
+    self.Schedule.On(self.DateRules.EveryDay('SPY'), self.TimeRules.At(18, 0), Action(self.Enter))
+    self.Schedule.On(self.DateRules.EveryDay('SPY'), self.TimeRules.At(5, 30), Action(self.Exit))
+    
+    # Afternoon session
+    self.Schedule.On(self.DateRules.EveryDay('SPY'), self.TimeRules.At(13, 15), Action(self.Enter))
+    self.Schedule.On(self.DateRules.EveryDay('SPY'), self.TimeRules.At(14, 0), Action(self.ExitException))
+    
+    # Close session
+    self.Schedule.On(self.DateRules.EveryDay('SPY'), self.TimeRules.At(14, 0), Action(self.EnterException))
+    self.Schedule.On(self.DateRules.EveryDay('SPY'), self.TimeRules.At(15, 15), Action(self.Exit))
+
+</pre>
+
+</div>
+
+<h3>Step 2:  Set up the Trading Signals</h3>
+
+<p>
+	For Combined Strategy, in Bullish Season, we buy the contract at 13:15, sell at 15:15 ( Afternoon Session + Close Session) and then buy again at 18:00 and sell it at 5:30 the next day (Night Session); in Bearish Season, we buy the contract at 13:15, sell at 14:00 ( Afternoon Session only) and then buy again at 18:00 and sell it at 5:30 the next day (Night Session). So there are two trades for every trading day. To realize this, we just need to check the month of the trading time and judge when to close the first trade (at 15:15 or at 14:00). 
+</p>
+
+<p>
+	For Night Session Strategy in Bullish Season or Bearish Season, we only buy at 18:00 and sell at 5:30 the next day, so there is only one trade for every trading day. This can also be realized by similar ideas (See code snippet for details).
+</p>
+
+<div class="section-example-container">
+
+<pre class="python">
+    def OnData(self, data):
+        '''OnData event is the primary entry point for your algorithm. Each new data point will be pumped in here.
+            Arguments:
+                data: Slice object keyed by symbol containing the stock data
+        '''
+        # self.Debug(f'Month: {self.Time.date().month}')
+        self.data = data
+        pass
+    
+    def Enter(self):
+        if not self.Portfolio.Invested:
+            
+            for chain in self.data.FutureChains:
+                contracts = list(filter(lambda x: x.Expiry < self.Time + timedelta(90), chain.Value))
+                
+                # self.Debug(f'Num of Contracts {self.Time}: {len(contracts)}')
+                
+                if len(contracts) == 0:
+                    continue
+                
+                front = sorted(contracts, key = lambda x: x.Expiry, reverse = False)[0]
+
+                # long
+                self.MarketOrder(front.Symbol, 1)
+    
+    def Exit(self):
+        if self.Portfolio.Invested:
+            # short
```

Comment:
Not technically "short" since we never sell a contract before we buy it. Change it to "liquidate"